File: js/src/crypto/encryption.ts
/**
 * Gitnix SDK - Encryption Layer
 *
 * Zero-knowledge client-side encryption using:
 * - XChaCha20-Poly1305 (via tweetnacl secretbox)
 * - Per-record random nonces (24 bytes)
 * - Fixed-size padding to prevent size analysis
 *
 * Note: tweetnacl's secretbox uses XSalsa20-Poly1305 which provides
 * equivalent security guarantees (256-bit key, 192-bit nonce, AEAD).
 */

import nacl from 'tweetnacl';
import { decodeUTF8, encodeUTF8, encodeBase64, decodeBase64 } from 'tweetnacl-util';
import type { EncryptedBlob, EncryptionConfig } from '../types/index.js';
import { GitnixError, GitnixErrorCode } from '../types/index.js';

/** Current encryption format version */
const BLOB_VERSION = 1;

/** Key size in bytes (256-bit) */
const KEY_SIZE = 32;

/** Nonce size in bytes (192-bit for XSalsa20) */
const NONCE_SIZE = 24;

/** Auth tag size (Poly1305) */
const AUTH_TAG_SIZE = 16;

export class Encryption {
  private config: Required<EncryptionConfig>;

  constructor(config: EncryptionConfig = {}) {
    this.config = {
      argon2MemoryCost: config.argon2MemoryCost ?? 65536,
      argon2TimeCost: config.argon2TimeCost ?? 3,
      argon2Parallelism: config.argon2Parallelism ?? 1,
      enablePadding: config.enablePadding ?? true,
      paddingBlockSize: config.paddingBlockSize ?? 256,
    };
  }

  // ─── Core Encrypt/Decrypt ──────────────────────────────────────────────

  /**
   * Encrypt plaintext bytes with a key.
   * Output format: [version(1)] [nonce(24)] [ciphertext+tag(N+16)]
   */
  encrypt(plaintext: Uint8Array, key: Uint8Array): Uint8Array {
    if (key.length !== KEY_SIZE) {
      throw new GitnixError(
        `Invalid key size: expected ${KEY_SIZE}, got ${key.length}`,
        GitnixErrorCode.ENCRYPTION_FAILED,
      );
    }

    // Apply padding if enabled
    const padded = this.config.enablePadding ? this.pad(plaintext) : plaintext;

    // Generate random nonce
    const nonce = nacl.randomBytes(NONCE_SIZE);

    // Encrypt with authenticated encryption (secretbox = XSalsa20-Poly1305)
    const ciphertext = nacl.secretbox(padded, nonce, key);

    // Assemble blob: [version][nonce][ciphertext]
    const blob = new Uint8Array(1 + NONCE_SIZE + ciphertext.length);
    blob[0] = BLOB_VERSION;
    blob.set(nonce, 1);
    blob.set(ciphertext, 1 + NONCE_SIZE);

    return blob;
  }

  /**
   * Decrypt an encrypted blob with a key.
   */
  decrypt(blob: Uint8Array, key: Uint8Array): Uint8Array {
    if (key.length !== KEY_SIZE) {
      throw new GitnixError(
        `Invalid key size: expected ${KEY_SIZE}, got ${key.length}`,
        GitnixErrorCode.DECRYPTION_FAILED,
      );
    }

    if (blob.length < 1 + NONCE_SIZE + AUTH_TAG_SIZE) {
      throw new GitnixError(
        'Encrypted blob too short',
        GitnixErrorCode.DECRYPTION_FAILED,
      );
    }

    const version = blob[0];
    if (version !== BLOB_VERSION) {
      throw new GitnixError(
        `Unsupported blob version: ${version}`,
        GitnixErrorCode.DECRYPTION_FAILED,
      );
    }

    const nonce = blob.slice(1, 1 + NONCE_SIZE);
    const ciphertext = blob.slice(1 + NONCE_SIZE);

    // Decrypt and verify authentication tag
    const decrypted = nacl.secretbox.open(ciphertext, nonce, key);
    if (decrypted === null) {
      throw new GitnixError(
        'Decryption failed: invalid key or tampered data',
        GitnixErrorCode.DECRYPTION_FAILED,
      );
    }

    // Remove padding if enabled
    return this.config.enablePadding ? this.unpad(decrypted) : decrypted;
  }

  // ─── JSON Helpers ──────────────────────────────────────────────────────

  /**
   * Encrypt a JSON-serializable object
   */
  encryptJSON(data: unknown, key: Uint8Array): Uint8Array {
    const json = JSON.stringify(data);
    const bytes = decodeUTF8(json);
    return this.encrypt(bytes, key);
  }

  /**
   * Decrypt to a JSON object
   */
  decryptJSON<T = unknown>(blob: Uint8Array, key: Uint8Array): T {
    const decrypted = this.decrypt(blob, key);
    const json = encodeUTF8(decrypted);
    try {
      return JSON.parse(json) as T;
    } catch {
      throw new GitnixError(
        'Failed to parse decrypted data as JSON',
        GitnixErrorCode.DECRYPTION_FAILED,
      );
    }
  }

  // ─── Base64 Helpers (for GitHub API) ───────────────────────────────────

  /**
   * Encrypt and return as base64 (ready for GitHub blob API)
   */
  encryptToBase64(plaintext: Uint8Array, key: Uint8Array): string {
    const encrypted = this.encrypt(plaintext, key);
    return encodeBase64(encrypted);
  }

  /**
   * Decrypt from base64 string
   */
  decryptFromBase64(base64: string, key: Uint8Array): Uint8Array {
    const blob = decodeBase64(base64);
    return this.decrypt(blob, key);
  }

  /**
   * Encrypt JSON and return as base64
   */
  encryptJSONToBase64(data: unknown, key: Uint8Array): string {
    const encrypted = this.encryptJSON(data, key);
    return encodeBase64(encrypted);
  }

  /**
   * Decrypt base64 to JSON
   */
  decryptJSONFromBase64<T = unknown>(base64: string, key: Uint8Array): T {
    const blob = decodeBase64(base64);
    return this.decryptJSON<T>(blob, key);
  }

  // ─── Key Operations ────────────────────────────────────────────────────

  /**
   * Generate a random 256-bit key
   */
  generateKey(): Uint8Array {
    return nacl.randomBytes(KEY_SIZE);
  }

  /**
   * Encrypt a key with the master key (key wrapping)
   */
  wrapKey(key: Uint8Array, masterKey: Uint8Array): { encrypted: Uint8Array; nonce: Uint8Array } {
    const nonce = nacl.randomBytes(NONCE_SIZE);
    const encrypted = nacl.secretbox(key, nonce, masterKey);
    return { encrypted, nonce };
  }

  /**
   * Decrypt a wrapped key
   */
  unwrapKey(encryptedKey: Uint8Array, nonce: Uint8Array, masterKey: Uint8Array): Uint8Array {
    const key = nacl.secretbox.open(encryptedKey, nonce, masterKey);
    if (key === null) {
      throw new GitnixError(
        'Failed to unwrap key: invalid master key',
        GitnixErrorCode.INVALID_PASSWORD,
      );
    }
    return key;
  }

  /**
   * Create a verification hash to validate the master password
   * without storing the key itself
   */
  createVerificationHash(masterKey: Uint8Array): Uint8Array {
    // Hash a known constant with the key
    const constant = decodeUTF8('GITNIX_KEY_VERIFICATION_v1');
    const nonce = new Uint8Array(NONCE_SIZE); // Zero nonce is OK here (deterministic)
    return nacl.secretbox(constant, nonce, masterKey);
  }

  /**
   * Verify a master key against a stored verification hash
   */
  verifyMasterKey(masterKey: Uint8Array, verificationHash: Uint8Array): boolean {
    const constant = decodeUTF8('GITNIX_KEY_VERIFICATION_v1');
    const nonce = new Uint8Array(NONCE_SIZE);
    const result = nacl.secretbox.open(verificationHash, nonce, masterKey);
    return result !== null;
  }

  // ─── Hashing ───────────────────────────────────────────────────────────

  /**
   * SHA-256 hash for content addressing and filename obfuscation
   */
  hash(data: Uint8Array): Uint8Array {
    return nacl.hash(data).slice(0, 32); // nacl.hash is SHA-512, take first 32 bytes
  }

  /**
   * Hash a string (for filenames, collection names)
   */
  hashString(str: string): string {
    const bytes = decodeUTF8(str);
    const hashed = this.hash(bytes);
    return encodeBase64(hashed)
      .replace(/\+/g, '-')
      .replace(/\//g, '_')
      .replace(/=+$/, '');
  }

  // ─── Padding ───────────────────────────────────────────────────────────

  /**
   * PKCS7-style padding to fixed block size.
   * Hides the true size of records from traffic analysis.
   */
  private pad(data: Uint8Array): Uint8Array {
    const blockSize = this.config.paddingBlockSize;
    // Calculate padded size: next multiple of blockSize
    const paddedSize = Math.ceil((data.length + 4) / blockSize) * blockSize;
    const padded = new Uint8Array(paddedSize);

    // Store original length in first 4 bytes (big-endian uint32)
    const view = new DataView(padded.buffer);
    view.setUint32(0, data.length, false);

    // Copy data after length prefix
    padded.set(data, 4);

    // Fill remaining with random bytes (not zeros, to avoid patterns)
    const randomFill = nacl.randomBytes(paddedSize - 4 - data.length);
    padded.set(randomFill, 4 + data.length);

    return padded;
  }

  /**
   * Remove padding
   */
  private unpad(data: Uint8Array): Uint8Array {
    if (data.length < 4) {
      throw new GitnixError('Invalid padded data', GitnixErrorCode.DECRYPTION_FAILED);
    }

    const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
    const originalLength = view.getUint32(0, false);

    if (originalLength > data.length - 4) {
      throw new GitnixError(
        'Invalid padding: claimed length exceeds data',
        GitnixErrorCode.DECRYPTION_FAILED,
      );
    }

    return data.slice(4, 4 + originalLength);
  }

  // ─── Config Access ─────────────────────────────────────────────────────

  getConfig(): Readonly<Required<EncryptionConfig>> {
    return { ...this.config };
  }
}

File: python/gitnix/encryption.py
"""Gitnix SDK - Encryption Layer.

Zero-knowledge client-side encryption using:
- XSalsa20-Poly1305 (via PyNaCl SecretBox)
- Per-record random nonces (24 bytes)
- Fixed-size padding to prevent size analysis
"""

from __future__ import annotations

import json
import struct
from base64 import b64decode, b64encode
from hashlib import sha512
from typing import Any

import nacl.secret
import nacl.utils
from nacl.exceptions import CryptoError

from gitnix.errors import GitnixError, GitnixErrorCode
from gitnix.types import EncryptionConfig

# Constants
BLOB_VERSION = 1
KEY_SIZE = 32
NONCE_SIZE = 24


class Encryption:
    """Handles all encryption/decryption operations."""

    def __init__(self, config: EncryptionConfig | None = None) -> None:
        self._config = config or EncryptionConfig()

    # ─── Core Encrypt/Decrypt ────────────────────────────────────────────

    def encrypt(self, plaintext: bytes, key: bytes) -> bytes:
        """Encrypt plaintext with key. Returns: version(1) + nonce(24) + ciphertext."""
        if len(key) != KEY_SIZE:
            raise GitnixError(
                f"Invalid key size: expected {KEY_SIZE}, got {len(key)}",
                GitnixErrorCode.ENCRYPTION_FAILED,
            )

        padded = self._pad(plaintext) if self._config.enable_padding else plaintext
        box = nacl.secret.SecretBox(key)
        encrypted = box.encrypt(padded)  # nonce + ciphertext

        # Format: [version(1)] [nonce(24)] [ciphertext(N)]
        return bytes([BLOB_VERSION]) + encrypted

    def decrypt(self, blob: bytes, key: bytes) -> bytes:
        """Decrypt an encrypted blob."""
        if len(key) != KEY_SIZE:
            raise GitnixError(
                f"Invalid key size: expected {KEY_SIZE}, got {len(key)}",
                GitnixErrorCode.DECRYPTION_FAILED,
            )

        if len(blob) < 1 + NONCE_SIZE + 16:
            raise GitnixError("Encrypted blob too short", GitnixErrorCode.DECRYPTION_FAILED)

        version = blob[0]
        if version != BLOB_VERSION:
            raise GitnixError(
                f"Unsupported blob version: {version}",
                GitnixErrorCode.DECRYPTION_FAILED,
            )

        # The rest is nonce + ciphertext (PyNaCl format)
        encrypted = blob[1:]
        box = nacl.secret.SecretBox(key)

        try:
            decrypted = box.decrypt(encrypted)
        except CryptoError:
            raise GitnixError(
                "Decryption failed: invalid key or tampered data",
                GitnixErrorCode.DECRYPTION_FAILED,
            )

        return self._unpad(decrypted) if self._config.enable_padding else decrypted

    # ─── JSON Helpers ────────────────────────────────────────────────────

    def encrypt_json(self, data: Any, key: bytes) -> bytes:
        """Encrypt a JSON-serializable object."""
        plaintext = json.dumps(data, separators=(",", ":")).encode("utf-8")
        return self.encrypt(plaintext, key)

    def decrypt_json(self, blob: bytes, key: bytes) -> Any:
        """Decrypt to a JSON object."""
        decrypted = self.decrypt(blob, key)
        try:
            return json.loads(decrypted.decode("utf-8"))
        except (json.JSONDecodeError, UnicodeDecodeError):
            raise GitnixError(
                "Failed to parse decrypted data as JSON",
                GitnixErrorCode.DECRYPTION_FAILED,
            )

    # ─── Base64 Helpers ──────────────────────────────────────────────────

    def encrypt_to_base64(self, plaintext: bytes, key: bytes) -> str:
        """Encrypt and return as base64."""
        return b64encode(self.encrypt(plaintext, key)).decode("ascii")

    def decrypt_from_base64(self, b64: str, key: bytes) -> bytes:
        """Decrypt from base64 string."""
        return self.decrypt(b64decode(b64), key)

    def encrypt_json_to_base64(self, data: Any, key: bytes) -> str:
        """Encrypt JSON and return as base64."""
        return b64encode(self.encrypt_json(data, key)).decode("ascii")

    def decrypt_json_from_base64(self, b64: str, key: bytes) -> Any:
        """Decrypt base64 to JSON."""
        return self.decrypt_json(b64decode(b64), key)

    # ─── Key Operations ──────────────────────────────────────────────────

    def generate_key(self) -> bytes:
        """Generate a random 256-bit key."""
        return nacl.utils.random(KEY_SIZE)

    def wrap_key(self, key: bytes, master_key: bytes) -> tuple[bytes, bytes]:
        """Encrypt a key with the master key. Returns (encrypted, nonce)."""
        box = nacl.secret.SecretBox(master_key)
        encrypted = box.encrypt(key)
        # PyNaCl prepends nonce to ciphertext
        nonce = encrypted[:NONCE_SIZE]
        ciphertext = encrypted[NONCE_SIZE:]
        return ciphertext, nonce

    def unwrap_key(self, encrypted_key: bytes, nonce: bytes, master_key: bytes) -> bytes:
        """Decrypt a wrapped key."""
        box = nacl.secret.SecretBox(master_key)
        try:
            return box.decrypt(encrypted_key, nonce)
        except CryptoError:
            raise GitnixError(
                "Failed to unwrap key: invalid master key",
                GitnixErrorCode.INVALID_PASSWORD,
            )

    def create_verification_hash(self, master_key: bytes) -> bytes:
        """Create a hash to verify the master password."""
        constant = b"GITNIX_KEY_VERIFICATION_v1"
        box = nacl.secret.SecretBox(master_key)
        nonce = b"\x00" * NONCE_SIZE
        return box.encrypt(constant, nonce)[NONCE_SIZE:]  # Skip prepended nonce

    def verify_master_key(self, master_key: bytes, verification_hash: bytes) -> bool:
        """Verify a master key against stored hash."""
        constant = b"GITNIX_KEY_VERIFICATION_v1"
        box = nacl.secret.SecretBox(master_key)
        nonce = b"\x00" * NONCE_SIZE
        try:
            result = box.decrypt(verification_hash, nonce)
            return result == constant
        except CryptoError:
            return False

    # ─── Hashing ─────────────────────────────────────────────────────────

    def hash_string(self, s: str) -> str:
        """Hash a string for filename obfuscation (URL-safe base64)."""
        digest = sha512(s.encode("utf-8")).digest()[:32]
        return b64encode(digest).decode("ascii").replace("+", "-").replace("/", "_").rstrip("=")

    # ─── Padding ─────────────────────────────────────────────────────────

    def _pad(self, data: bytes) -> bytes:
        """PKCS7-style padding to fixed block size."""
        block_size = self._config.padding_block_size
        padded_size = ((len(data) + 4 + block_size - 1) // block_size) * block_size
        length_prefix = struct.pack(">I", len(data))
        padding = nacl.utils.random(padded_size - 4 - len(data))
        return length_prefix + data + padding

    def _unpad(self, data: bytes) -> bytes:
        """Remove padding."""
        if len(data) < 4:
            raise GitnixError("Invalid padded data", GitnixErrorCode.DECRYPTION_FAILED)
        original_length = struct.unpack(">I", data[:4])[0]
        if original_length > len(data) - 4:
            raise GitnixError(
                "Invalid padding: claimed length exceeds data",
                GitnixErrorCode.DECRYPTION_FAILED,
            )
        return data[4 : 4 + original_length]
